### PR TITLE
Fix interface conversion in document

### LINKED ISSUE #53: panic: interface conversion: interface {} is nil, not *common.ElementHandle
The following script causes `panic: interface conversion: interface {} is nil, not *common.ElementHandle` during the call to `page.waitForSelector("li.design-tile");`:

```javascript
import { sleep, group } from 'k6';
import launcher from 'k6/x/browser';

import { randomIntBetween } from 'https://jslib.k6.io/k6-utils/1.1.0/index.js';

export const options = {};

let browser, context, page;

const pauseMin = 5;
const pauseMax = 5;

export default function() {
  initialize();

  navigateToHomepage();
  clickBusinessCards();
  clickBrowseDesigns();
  clickDesign('Standard (3.5" x 2")');
}

function initialize() {
  browser = launcher.launch('chromium', {
    args: [],
    debug: false,
    devtools: false,
    env: {},
    executablePath: null,
    headless: false,
    ignoreDefaultArgs: [],
    proxy: {},
    slowMo: '500ms',
    timeout: '30s',
  });

  context = browser.newContext({
    ignoreHTTPSErrors: true
  });

  page = context.newPage();
}

function navigateToHomepage() {
  console.log('Starting navigateToHomepage...');

  group("Navigate to Homepage", () => {
    page.goto('https://vistaprint.ca', { waitUntil: 'networkidle' });

    page.waitForSelector(".standard-hero-text");
  });

  sleep(randomIntBetween(pauseMin, pauseMax));
}

function clickBusinessCards() {
  console.log('Starting clickBusinessCards...');

  const elem = page.$("//span[text()='Business Cards']");

  group("Click Business Cards", () => {
    elem.click();
    
    // Wait for next page to load
    // page.waitForLoadState('networkidle');

    // Wait for a specific element to appear
    page.waitForSelector("//a[text()='Browse designs']");
  });

  sleep(randomIntBetween(pauseMin, pauseMax));
}

function clickBrowseDesigns() {
  console.log('Starting clickBrowseDesigns...');

  const elem = page.$("//a[text()='Browse designs']");

  group("Click Browse Designs", () => {
    elem.click();
    
    // page.waitForLoadState('networkidle');

    page.waitForSelector("li.design-tile");
  });

  sleep(randomIntBetween(pauseMin, pauseMax));
}

function clickDesign(designName) {
  console.log('Starting clickDesign...');
  
  // fetch all of the design options
  const designOptions = page.$("li.design-tile");

  // ... incomplete script ;-)
}
```

Uncommenting `page.waitForLoadState('networkidle')` allows the script to proceed (although it eventually fails with `panic: send on closed channel` as per #49).

The only exception I'm expecting here is a timeout in case the selector was unsuccessful.

I think what might be happening is that the auto-wait on `elem.click();` is not waiting long enough for the page to even change (the site is very slow at switching), and so we attempt to `page.waitForSelector("li.design-tile")` on a page/element that's about to be "detached".

### PR DESCRIPTION
**Pull request by inancgumus**
The error was:
panic: interface conversion: interface {} is nil, not
*common.ElementHandle

The cause of the error was trying to assert a nil result
in Frame.document to *ElementHandle. Of course, this is
probably the surface error. There may be other causes
for it, or maybe not.

It was hard to test Frame, so I used an abstraction called
FrameExecutionContext. There are a lot of problems are
happening with execution contexts. See: #125 and #49, for
example. So, it's better to make them testable, starting
here.

Fixes #53

### CODE REVIEW

File: common/frame_test.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	// frame should not panic with a nil document
+	stub := &executionContextTestStub{
+		evaluateFn: func(apiCtx context.Context, forceCallable bool, returnByValue bool, pageFunc goja.Value, args ...goja.Value) (res interface{}, err error) {
+			// return nil to test for panic
+			return nil, nil
+		},
+	}
+
+	// document() waits for the main execution context
+	go frame.setContext("main", stub)
```
Comment (imiric) [inline common/frame_test.go:48]: Let's wait with a timeout (that fails the test) for this goroutine to finish. The fact it does is an implementation detail, and the test should be more robust.

**Review by imiric — CHANGES_REQUESTED** [review]
LGTM, just minor suggestions.

While this fixes the panic, I'm not sure what effects it will functionally have for all callers of `Frame.document()`. I guess we'll find out and deal with that if/when it happens.

Have you tested how this behaves with the script in #53?

File: common/frame_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"context"
+	"testing"
+
+	"github.com/chromedp/cdproto/cdp"
+	"github.com/dop251/goja"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFrameDocument_Issue53(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/frame_test.go:32]: `Issue53` tells people to look for more information if any regression occurs. You can see the same convention in the Go source code, for example:

https://github.com/golang/go/blob/d6c4583ad4923533ddc9f5792ed3b66f3b9f9feb/src/go/printer/printer_test.go#L737-L758

Or:

https://github.com/golang/go/blob/d6c4583ad4923533ddc9f5792ed3b66f3b9f9feb/src/go/types/api_test.go#L1714-L1729

But I'll describe what it solves and remove underscores, np.

File: common/frame_execution_context.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	"context"
+
+	"github.com/chromedp/cdproto/cdp"
+	"github.com/chromedp/cdproto/runtime"
+	"github.com/dop251/goja"
+	"github.com/grafana/xk6-browser/api"
+)
+
+// frameExecutionContext represents a JS execution context that belongs to Frame.
+type frameExecutionContext interface {
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/frame_execution_context.go:33]: No, we do not 😄

File: common/frame_test.go (hunk 1 above)
Comment (inancgumus) [reply common/frame_test.go:48]: Hmm, yeah, good call.

File: common/frame_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (imiric) [reply common/frame_test.go:32]: That's great for Go core developers, we don't need to blindly follow whatever they do. `Issue53` is obscure and requires a) looking up the issue in the first place, b) access and dependency on GitHub issues.

Feel free to add a link to the GH issue as a comment, even though the "Fixes #53" reference in the commit message would be enough for anyone who wants to look up further information.

**Comment by inancgumus** [comment]
> LGTM, just minor suggestions.

Thanks 🙏

> While this fixes the panic, I'm not sure what effects it will functionally have for all callers of `Frame.document()`. I guess we'll find out and deal with that if/when it happens.

Sadly, yes, we'll find that out as we improve the tests.

> Have you tested how this behaves with the script in #53?

Sure thing.

**Review by imiric — APPROVED** [review]
LGTM

Maybe update the PR description with how it behaves with the script from #53, or comment in the issue itself.

Also, the `bug` label is typically only useful on issues, unless your PR also introduces a bug :stuck_out_tongue:

**Comment by inancgumus** [comment]
> Maybe update the PR description with how it behaves with the script from #53, or comment in the issue itself.

It behaves like without the error 😄  More like other unsolvable errors we've seen so far.

File: common/frame_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/frame_test.go]: Thousands of people contribute to Go, not only Go core devs, and the code is diligently reviewed and refactored according to idiomatic Go code standards (I know it from my experience). Anyway, I already updated it cause I slightly agree :)

**Pull request merged**